### ISSUE TITLE
re-train existing model with new params

### ISSUE THREAD

**Issue by tabee** [tabee/Reinforcement-Learning-Snake#29]
Ja, du kannst sowohl **die Rewards** als auch **die Modellparameter** ändern, wenn du ein geladenes Modell weitertrainierst! Allerdings gibt es einige Dinge zu beachten.

---

## **1. Rewards ändern beim Weitertraining**
✅ Du kannst die **Reward-Struktur** jederzeit in der Gym-Umgebung ändern, bevor du das Modell weitertrainierst.  

### **So geht’s:**
Falls deine `SnakeEnv`-Umgebung so aufgebaut ist:
```python
class SnakeEnv(gym.Env):
    def __init__(self):
        ...
        self.reward_for_food = 1.0
        self.reward_for_survival = 0.01
        self.penalty_for_collision = -1.0

    def step(self, action):
        if self.eats_food():
            reward = self.reward_for_food
        elif self.hits_wall() or self.bites_itself():
            reward = self.penalty_for_collision
        else:
            reward = self.reward_for_survival
        return obs, reward, done, info
```
Dann kannst du das Modell **nach dem Laden mit neuen Rewards weitertrainieren**:

```python
model = PPO.load("./best_model/best_model.zip", env=env)

# Ändere die Reward-Struktur der Umgebung
env.reward_for_food = 2.0  # Doppelte Belohnung für Essen
env.reward_for_survival = 0.02  # Höhere Belohnung fürs Überleben

# Setze das Training fort
model.learn(total_timesteps=200000)
```
📌 **Wichtig:**  
- Das neuronale Netz wird **nicht zurückgesetzt** – es trainiert weiter mit den **neuen Rewards**.  
- Das kann dazu führen, dass sich das Modell an die neuen Belohnungen **schneller oder langsamer** anpasst.  

---

## **2. Modellparameter ändern beim Weitertraining**
Ja, du kannst einige Modellparameter **ändern**, aber **nicht alle**.  

### **Welche Parameter kannst du ändern?**
✅ **Diese Parameter kannst du beim Laden ändern:**
- `learning_rate`: Kann **on-the-fly** angepasst werden, um feineres Training zu ermöglichen.  
- `clip_range`: Kann geändert werden, um aggressivere oder sanftere Policy-Updates zu steuern.  
- `ent_coef`: Falls dein Modell nicht genug Exploration betreibt, kannst du dies erhöhen.  

### **Welche Parameter kannst du NICHT ändern?**
🚫 Die Architektur des neuronalen Netzes (`policy`) ist **festgelegt**. Falls du z. B. ein `MlpPolicy`-Modell trainiert hast, kannst du es nicht direkt zu einem `CnnPolicy` ändern.  

---

### **So änderst du Modellparameter beim Weitertrainieren**
Du kannst diese Parameter ändern, indem du `policy_kwargs` nutzt:
```python
# Modell laden
model = PPO.load("./best_model/best_model.zip", env=env)

# Learning Rate anpassen (z. B. für feineres Training)
model.learning_rate = 0.0001  # Von 0.0003 auf 0.0001 reduziert

# Exploration anpassen
model.ent_coef = 0.02  # Mehr Exploration

# Training fortsetzen
model.learn(total_timesteps=200000)
```
📌 **Falls du größere Änderungen an der Architektur machen willst, musst du ein neues Modell erstellen.**  

---

## **3. Alternativer Trick: Modellparameter über Fine-Tuning ändern**
Falls du mehr Kontrolle brauchst, kannst du die **Gewichte in ein neues Modell übertragen**, aber mit neuen Hyperparametern.

```python
old_model = PPO.load("./best_model/best_model.zip", env=env)

# Neues Modell mit geänderten Parametern
new_model = PPO(
    "MlpPolicy",
    env,
    learning_rate=0.0001,  # Geänderte Lernrate
    ent_coef=0.02,  # Mehr Exploration
    clip_range=0.1,  # Weniger aggressive Updates
    verbose=1
)

# Kopiere die Gewichte ins neue Modell
new_model.policy.load_state_dict(old_model.policy.state_dict())

# Training fortsetzen
new_model.learn(total_timesteps=200000)
```
📌 **Vorteil:**  
- Du kannst **beliebige Hyperparameter anpassen**, ohne das gesamte Modell neu zu trainieren.  

---

## **Fazit**
✅ **Ja, du kannst die Rewards und einige Parameter ändern**, während du mit dem Training fortfährst.  
✅ Die **beste Methode für größere Änderungen** ist **Fine-Tuning mit `load_state_dict()`**.  

Falls du spezielle Änderungen vorhast, kann ich dir helfen, den besten Ansatz zu finden! 😊
